### ISSUE TITLE
Investigate if we can do without networkx

### ISSUE THREAD

**Issue by flying-sheep** [scverse/scanpy#97]
AFAIK networkx and python-igraph do the same thing, only that python-igraph is faster.

We also need python-igraph anyway for louvain and so on, so maybe it would be good to get rid of networkx.

Downside: python-igraph and louvain-igraph is currently deliberately an optional dependency since it’s hard to install on windows. People need to build it themselves (A task that even I didn’t manage by now, and I got *many* things to compile!) or use Christoph Grohlke’s unofficial builds ([here](https://www.lfd.uci.edu/~gohlke/pythonlibs/#python-igraph) and [here](https://www.lfd.uci.edu/~gohlke/pythonlibs/#louvain-igraph))

**Comment by falexwolf** [comment]
Advantage of networkx is that it's easily installed... But yes, we should remove it in the future.

I think with anaconda, one gets all the igraph and louvain stuff to work very easily without compiling. Without using Grohlke's binaries... One just needs to document this probably.

At the latest when igraph and louvain are easily installed, networkx can be removed...

PS: I'm currently preparing scanpy 1.0; there will be some slight changes to make the API less redundant... So for now, please no big changes...

**Comment by flying-sheep** [comment]
oh, sure, i wouldn’t do big stuff without checking back with you first

**Comment by dawe** [comment]
Graph_tool library would be even better and it also implements SBM and many other things that may be useful in graph analysis of single cells. Unfortunately its installation is a painful experience.

**Comment by falexwolf** [comment]
Yes, graph_tool is nice and I'm also using it;  but yes, it's installation is even worse than igraph... hence, no option for a dependency...

**Comment by flying-sheep** [comment]
igraph wouldn’t be a problem if there were wheels on windows and macOS. i hope that comes one day

**Comment by dawe** [comment]
@flying-sheep [community module](http://python-louvain.readthedocs.io/en/latest/index.html), which builds on networkx, has a more than decent implementation of Louvain's method. Also, NetworkX 2.1 has been improved over its predecessors. Why dropping igraph isn't an option?

**Comment by flying-sheep** [comment]
@falexwolf correct me if i’m wrong, but: i think igraph is just super fast. most of our advertised speed comes from using it.

**Comment by falexwolf** [comment]
@dawe In all benchmarks that I did maybe a bit less than a year ago, the `python-louvain` didn't seem to produce satisfying results... But maybe I did something stupid. I'll reevaluate this, thanks, Davide! PS: One can easily switch between implementations; simply pass `flavor='taynaud'` to `sc.tl.louvain` and you'll use `python-louvain`. See [here](https://github.com/theislab/scanpy/blob/5299c6caaec6402513f1e0442186350787177d2c/scanpy/tools/louvain.py#L118-L125). However, I removed this from the docs as I was not so satisfied with it...

@flying-sheep the only thing where `igraph` is used in Scanpy is for graph drawing, where it's incredibly faster than `networkx` (completely forget about `networkx` in this respect); the performant `louvain` implementation is due to the `louvain` package, which simply uses `igraph`'s graph data structures

**Comment by dawe** [comment]
Hello again, some months later. Unfortunately I haven’t used scanpy for a while and now I’m back I found that only two flavors are available for Louvain processing, `igraph` and `vtraag`. The latter allows for resolution parameter but relies on `python-louvain`, right? the reason for my “but” is that I’m having issues with that module on OSX, which I understand it is not a specific scanpy issue. I’ll ask again: what’s wrong with `networkx` and `community` modules? I’m asking because flavor `taynaud` is no more available

**Comment by falexwolf** [comment]
I think `flavor='taynaud'` should [still work](https://github.com/theislab/scanpy/blob/a9ee39f5152d167f1aeb784ffbdd0a6e3dc1409a/scanpy/tools/louvain.py#L142), but is really only meant as a last resort. AFAIK the networkx modules don't even provide a resolution parameter, but most importantly, they don't scale. I think I even was satisfied with convergence of the results and comparisons with, e.g., Seurat.

Scanpy can even be installed using bioconda, so there should be no problem with igraph installation these days. Did you checkout https://scanpy.readthedocs.io/en/latest/installation.html?

**Comment by dawe** [comment]
I’m quite sure it has a resolution parameter, but at this point I’m also quite sure I’m messing up with modules and dependencies, both in this thread and on my local installation...  about conda, I guess I’m one of the last around who hasn’t adopted it yet
